src/sentry/ingest/billing_metrics_consumer.py

```diff
@@ -108,33 +124,143 @@ def __init__(
+        self._messages_ready_since_last_commit: int = 0
+
+    def _get_billing_producer(self) -> KafkaProducer:
+        servers = _get_bootstrap_servers(topic=self._billing_topic.name, force_cluster=None)
+        return KafkaProducer(
+            build_kafka_configuration(default_config={}, bootstrap_servers=servers)
+        )
 
     def poll(self) -> None:
+        self._mark_commit_ready()
         if self._should_commit():
             self._bulk_commit()
 
+    def _mark_commit_ready(self):
+        """Removes completed futures at the beginning of
+        ongoing_billing_outcomes and adds them to ready_to_commit.
+
+        If a future has errored, log an error but add to ready_to_commit
+        anyways.
+        """
+        while self._ongoing_billing_outcomes:
+            produce_billing, metrics_msg = self._ongoing_billing_outcomes[0]
+            if not produce_billing.done():
+                break
+
+            if produce_billing.exception():
+                logger.error(
+                    "Async future failed in billing metrics consumer.",
+                    exc_info=produce_billing.exception(),
+                    extra={"offset": metrics_msg.offset},
+                )
+
+            self._ongoing_billing_outcomes.popleft()
+            self._ready_to_commit[metrics_msg.partition] = Position(
+                metrics_msg.next_offset, datetime.now()
+            )
+            self._messages_ready_since_last_commit += 1
+
+    def _should_commit(self) -> bool:
+        if not self._ready_to_commit:
+            return False
+        if self._messages_ready_since_last_commit >= self._max_batch_size:
+            return True
+        if self._last_commit_time + self._max_batch_time <= datetime.now():
+            return True
+        return False
+
+    def _bulk_commit(self) -> None:
+        """Commits and clears the ready to commit queue."""
+        if self._messages_ready_since_last_commit == 0:
+            return
+        self._commit(self._ready_to_commit)
+        self._clear_ready_queue()
+
+    def _clear_ready_queue(self) -> int:
+        """Clears the ready to commit queue and returns the amount of items dropped."""
+        amount_dropped = self._messages_ready_since_last_commit
+        self._ready_to_commit = {}
+        self._messages_ready_since_last_commit = 0
+        self._last_commit_time = datetime.now()
+        return amount_dropped
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        deadline = time.time() + timeout if timeout else None
+        self._bulk_commit()
+        while self._ongoing_billing_outcomes:
+            if deadline and deadline > time.time():
+                items_left = self._messages_ready_since_last_commit + len(
+                    self._ongoing_billing_outcomes
+                )
+                logger.warning(f"join timed out, items in the queue left: {items_left}")
+                break
+            self._mark_commit_ready()
```

Comment:
Other consumers call `Future.result()` at this point, which blocks until it finishes or the time runs out. `_mark_commit_ready()` does not block, so in its current form this `while` loop is a busy loop.

https://github.com/getsentry/sentry/blob/5f85f591c299c83a64264edc000daba8c8f77ed7/src/sentry/replays/consumers/recording/process_recording.py#L221